six/modules/c++/six.sicd/tests/test_add_additional_des.cpp

```diff
@@ -0,0 +1,272 @@
+* see <http://www.gnu.org/licenses/>.
+*
+*/
+
+#include <cstdio>
+#include <six/NITFWriteControl.h>
+#include <six/Types.h>
+#include <six/XMLControlFactory.h>
+#include <six/sicd/ComplexXMLControl.h>
+#include <six/sicd/Utilities.h>
+
+namespace
+{
+void validateArguments(int argc, char** argv)
+{
+    if (argc != 2)
+    {
+        std::string message = "Usage: " + sys::Path::basename(argv[0])
+                + " <XML pathname>";
+        throw except::Exception(Ctxt(message));
+    }
+    if (!str::endsWith(argv[1], ".xml"))
+    {
+        std::string message = "Extension suggests input may not be an XML file. "
+            "Expecting '*.xml'.";
+        throw except::Exception(Ctxt(message));
+    }
+}
+
+std::vector<sys::Int16_T> generateBandData(const six::sicd::ComplexData& data)
+{
+    std::vector<sys::Int16_T> bandData(data.getNumRows() * data.getNumCols());
+
+    for (size_t ii = 0; ii < bandData.size(); ++ii)
+    {
+        bandData[ii] = static_cast<sys::Int16_T>(ii);
+    }
+
+    return bandData;
+}
+
+bool addingNullSegmentWriterShouldThrow(const std::string& xmlPathname)
+{
+    std::cout << "Running addingNullSegmentWriterShouldThrow\n";
+    std::auto_ptr<six::sicd::ComplexData> data =
+            six::sicd::Utilities::parseDataFromFile(xmlPathname,
+            std::vector<std::string>(),
+            logging::NullLogger());
+
+    six::Container container(six::DataType::COMPLEX);
+    container.addData(data.release());
+
+    six::NITFWriteControl writer;
+    writer.initialize(&container);
+
+    mem::SharedPtr<nitf::SegmentWriter> segmentWriter;
+    try
+    {
+        writer.addAdditionalDES(segmentWriter);
+        // Above line should throw
+        std::cerr << "Test failed\n";
+        return false;
+    }
+    catch (const except::Exception&)
+    {
+        std::cout << "Test passed\n";
+        return true;
+    }
+}
+
+
+bool addingUnloadedSegmentWriterShouldThrow(const std::string& xmlPathname)
+{
+    std::cout << "Running addingUnLoadedSegmentWriterShouldThrow\n";
+    std::auto_ptr<six::sicd::ComplexData> data =
+            six::sicd::Utilities::parseDataFromFile(xmlPathname,
+            std::vector < std::string>(),
+            logging::NullLogger());
+    std::vector<sys::Int16_T> bandData(
+        generateBandData(*data));
+
+    six::Container container(six::DataType::COMPLEX);
+    container.addData(data.release());
+
+    six::NITFWriteControl writer;
+    writer.initialize(&container);
+
+    nitf::Record record = writer.getRecord();
+    nitf::DESegment des = record.newDataExtensionSegment();
+    des.getSubheader().getFilePartType().set("DE");
+    des.getSubheader().getTypeID().set("XML_DATA_CONTENT_005");
+    des.getSubheader().getVersion().set("01");
+    des.getSubheader().getSecurityClass().set("U");
+
+    mem::SharedPtr<nitf::SegmentWriter> segmentWriter(new nitf::SegmentWriter);
+    writer.addAdditionalDES(segmentWriter);
+
+    std::string name(std::tmpnam(NULL));
```

Comment:
What if the writer starts to write the file and then throws for some reason?  You won't ever call `std::remove()` here.  A simpler way to do this is to make an RAII object that at construction calls `tmpnam()`, at destruction calls `std::remove`, and has a method to give you the name of the file so you can pass it to `writer.save()`.